lib/crystal/benchmark.cr

```diff
@@ -0,0 +1,25 @@
+module Enumerable
+    def variance
+      mean = self.sum / self.size
+      sum = self.reduce(0) { |acc, cur| acc + (cur-mean) ** 2 }
+      sum/(self.size - 1).to_f
+    end
+
+    def std_dev
+      Math.sqrt(self.variance)
+    end
+end
+
+def bench(run_ms : Int32, warmup? : Bool = false, &fn)
+	times = Array(Int32).new
+	result = 0
+	while(times.sum < run_ms && !(times.sum == 0 && times.size > 0))
+		a = Time.measure {result = yield}
+		times << a.milliseconds
+	end
+
+	# mean_ms,std-dev-ms,min_ms,max_ms,times,result
+	unless times.empty? || warmup?
+		puts "#{times.sum / times.size},#{times.std_dev},#{times.min},#{times.max},#{times.size},#{result}"
+	end
+end
```

Comment:
There probably should be two functions in the interface here: `bench`/`run`, and `format_results`. Where the run function returns its results (the `times` array in this case) and `format_results,` well, formats a given result.

This way a run module like that for levenshtein can sum up the distances returned in the result and get help to format the output.

Reply:
There should be some status reporting here. We do it super simple with printing a dot every second. Se other benchmark modules for reference. (Clojure is most similar to Crystal of the current implementations, so may be the easiest one to port.)